### ISSUE TITLE
How to combine optimizing a model on a kubernetes cluster (using DB) and MlFlow?

### ISSUE THREAD

**Issue by konradsemsch** [optuna/optuna#1297]
Hello, could you please advise how to go about optimizing the model on a cluster and combining it with an Ml Flow dashboard across all combined runs?

As far as I understand, if I add this to my script in the docker image, the UI will be available for runs performed by a single pod, correct?
https://github.com/optuna/optuna/blob/master/examples/mlflow/keras_mlflow.py

I've also been looking at this example provided by Ml Flow ([link](https://www.mlflow.org/docs/latest/projects.html#run-an-mlflow-project-on-kubernetes-experimental)), but I'm not 100% sure how to connect the dots with optimization running on cluster. Could you please give some hints on that?

**Comment by g-votte** [comment]
There would be a variety of ways to run Optuna with Kubernetes and MLflow. A possible solution would be: based on the [Kubernetes example](https://github.com/optuna/optuna/tree/master/examples/distributed/kubernetes), modifying the manifest to launch a MLflow pod. You can use [`MLflowCallback`](https://optuna.readthedocs.io/en/stable/reference/integration.html#optuna.integration.MLflowCallback) by passing it to the `study.optimize` method specifying the tracking URI of the launched MLflow service.

> As far as I understand, if I add this to my script in the docker image, the UI will be available for runs performed by a single pod, correct?
> https://github.com/optuna/optuna/blob/master/examples/mlflow/keras_mlflow.py

Please note that the UI is not launched just by running the script. When you run it in your local environment, you’ll need to run the command of `mlflow ui` after running the study, as described in the [example’s README](https://github.com/optuna/optuna/tree/master/examples/mlflow). When using MLflow in Kubernetes, you might need to additionally call `mlflow.set_tracking_uri` in the script and pass the launched MLflow URI.

Please also note that the example you referred shows how to put Optuna data to MLflow without `MLflowCallback`.

**Comment by hvy** [comment]
https://github.com/optuna/optuna/pull/1464 demonstrates how one could go about it.

**Comment by konradsemsch** [comment]
That's very helpful! You're a little bit faster than as as we've almost managed to do it by ourselves 😉 we'll take your reference and complete our won implementation - thank you!

**Comment by hvy** [comment]
@konradsemsch, curious but did you manage to get it to work? We verified the toy example in #1464 and it has actually been merged into the master branch.

If you have some comments for future readers, that'd be great. I'm otherwise wondering if we should close this issue as fixed. Please let me know when you have time.

**Comment by konradsemsch** [comment]
Hi @hvy ! Let me update you again once my DevOps colleague who implemented it is back form holidays :)

**Comment by sachinruk** [comment]
If I understand [this example](https://github.com/optuna/optuna/tree/master/examples/kubernetes/simple) correctly it's running 5 pods in parallel with each pod running 20 iterations of optuna. However, the other pods can't see or use the iterations used in the other pods. Is that correct?
